llvm/lib/Analysis/CmpInstAnalysis.cpp

```diff
@@ -147,6 +150,19 @@ llvm::decomposeBitTestICmp(Value *LHS, Value *RHS, CmpInst::Predicate Pred,
 
     return std::nullopt;
   }
+  case ICmpInst::ICMP_EQ:
+  case ICmpInst::ICMP_NE: {
+    assert(DecomposeBitMask);
+    const APInt *AndC;
+    Value *AndVal;
+    if (match(LHS, m_And(m_Value(AndVal), m_APIntAllowPoison(AndC)))) {
+      Result = {AndVal /*X*/, Pred /*Pred*/, *AndC /*Mask*/, *OrigC /*C*/};
```

Comment:
```suggestion
      LHS = AndVal;
      Result.Mask = *AndC;
      Result.C = C;
      Result.Pred = Pred;
```
In this way, we can catch the pattern `(trunc(X) & Mask) eq/ne C`.